src/tools/rest-backend/service.hpp

```diff
@@ -0,0 +1,176 @@
+/**
+ * @file
+ *
+ * @brief header for the service classes
+ *
+ * @copyright BSD License (see doc/LICENSE.md or http://www.libelektra.org)
+ */
+
+#ifndef ELEKTRA_REST_SERVICE_HPP
+#define ELEKTRA_REST_SERVICE_HPP
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <boost/thread/locks.hpp>
+#include <boost/thread/shared_mutex.hpp>
+#include <cppcms/json.h>
+
+#include <config.hpp>
+#include <exceptions.hpp>
+#include <kdb_includes.hpp>
+#include <model_configformat.hpp>
+#include <model_entry.hpp>
+#include <model_importedconfig.hpp>
+#include <model_pluginformat.hpp>
+#include <model_user.hpp>
+#include <singleton.hpp>
+
+/**
+ * @brief main namespace for the REST service
+ */
+namespace kdbrest
+{
+
+/**
+ * @brief namespace for service classes
+ */
+namespace service
+{
+
+/**
+ * @brief service offering search and filter functionality
+ * 
+ * Service that is responsible for doing search specific tasks such as
+ * filtering a result set based on search parameters.
+ * 
+ * It is designed as singleton, so all applications using it will share
+ * the same instance, which allows for further logic if necessary.
+ */
+class SearchEngine : public singleton<SearchEngine>
+{
+
+public:
+	// configuration entries
+	void filterConfigurationsByName (std::vector<kdbrest::model::Entry> & entries, const std::string & startsWith);
+	void findConfigurationsByFilter (std::vector<kdbrest::model::Entry> & entries, const std::string & filter,
+					 const std::string filterby = Config::output_default_entry_filterby);
+
+	// user entries
+	void findUsersByFilter (std::vector<kdbrest::model::User> & users, const std::string & filter,
+				const std::string filterby = Config::output_default_user_filterby);
+
+private:
+};
+
+/**
+ * @brief service offering storage functionality
+ * 
+ * Service that is responsible for storing configuration and user
+ * entries in the kdb database. Besides storing it is also capable of
+ * updating and deleting of single entries, as well as looking up
+ * single and multiple entries at the same time.
+ * 
+ * It is designed as singleton, so all applications using it will share
+ * the same instance, which allows for further logic if necessary.
+ */
+class StorageEngine : public singleton<StorageEngine>
+{
+
+public:
+	// constructor
+	StorageEngine ();
+
+	// configuration entries
+	bool createEntry (model::Entry & entry);
+	bool updateEntry (model::Entry & entry);
+	bool deleteEntry (model::Entry & entry);
+
+	bool entryExists (const std::string & key);
```

Comment:
Will probably work now, but as soon as I start using a global KDB instance again (for cachefilter later), it will not work.